### PR TITLE
✨ Better edge case and schema migration handling

### PR DESCRIPTION
**Pull request by daaain**
I just got started on this but thought I should share.

Main bits:
 * [x] handle empty `e.errors` list
 * [x] estimate message size to try in halves if necessary
 * [x] try migrating table schema with an update before delete + recreate

### CODE REVIEW

File: target_bigquery.py [hunk 1]
```diff
@@ -355,7 +352,23 @@ def write_rows_to_bigquery(streams, emit_state_after_write=False):
...
+                        # While this is similar to the above, we want to avoid doing the json.dumps
+                        # if possible, so will only do it if we didn't get the expected errors
+                        elif len(json.dumps(fixed_rows)) > 10000000:
+                            insert_in_halves()
+                        elif (
+                            # Normally there's an `errors` attribute with a list of errors on the
+                            # exception object if it comes from Google, but sometimes it's empty or
+                            # we get a native exception which doesn't have the `errors` attribute
+                            next(iter(getattr(e, "errors", [{}])), {}).get("reason")
+                            in RETRYABLE_ERROR_CODES
```
Comment (jmstaley) [inline target_bigquery.py]: this isn't very readable, surely you can do something like
```
errors = getattr(e, "errors", [])
.
.
.
elif errors and errors[0].get(reason):
```
if errors is an empty list it evaluates to `False` and the execution stops, makes it a lot more readable

Comment (daaain) [reply target_bigquery.py]: Right, empty arrays are truthy in Javascript so I missed that trick!

File: target_bigquery.py [hunk 2]
```diff
@@ -355,7 +352,23 @@ def write_rows_to_bigquery(streams, emit_state_after_write=False):
...
-                        elif getattr(e, "errors", [{}])[0].get("reason") in RETRYABLE_ERROR_CODES:
+
+                        if (
+                            "payload size exceeds the limit" in error_string
+                            or "too many rows present" in error_string
+                        ):
+                            insert_in_halves()
+                        # While this is similar to the above, we want to avoid doing the json.dumps
+                        # if possible, so will only do it if we didn't get the expected errors
+                        elif len(json.dumps(fixed_rows)) > 10000000:
```
Comment (sachams) [inline target_bigquery.py]: Didn't we have a problem with this where we tried 5k rows but still hit an error as the rows were very wide?

Comment (daaain) [reply target_bigquery.py]: This is the condition for that case, it dumps the JSON and checks the length so we'd see if it's likely to be above the 10MB threshold.

**Comment by daaain** [comment]
I think this is GTG now, please have another look @sachams @jmstaley 

While there's a reworked, optional support for the nuclear, table delete / re-create option, I managed to make it work with almost all migration options (ie column addition, deletion) using table update and it will only fail on impossible to work around ones like column type change. I've done this by adding an option (and making it default) to set all columns to nullable in the BQ schema (so we can add columns which are non-nullable in Postgres) and carrying on to try inserting rows even when the table update fails (in the case of column deletion).

Also fixed a bug (and wrote a test for it) which I haven't noticed until now, that even in case of failed rows the state is sent back 🙀

**Review by jmstaley — APPROVED** [review]

**Pull request merged**